use-core/src/main/java/org/tzi/use/uml/ocl/expr/ExpObjOp.java

```diff
@@ -39,132 +34,121 @@
                                               "Type mismatch in argument `" + params.varDecl(i - 1).name() + 
                                               "'. Expected type `" + params.varDecl(i - 1).type() + 
                                               "', found `" + args[i].type() + "'.");
     }
-    
+
     public Value eval(EvalContext ctx) {
-    	if (isPre()) {
-    		ctx = new EvalContext(ctx.preState(), ctx.preState(), ctx.varBindings(), ctx);
-    	}
-    	
-    	ctx.enter(this);
-    	
-    	Value result = UndefinedValue.instance;
-    	
-    	Value selfVal = fArgs[0].eval(ctx);
-    	
-    	if (selfVal.isUndefined() || 
-    			!(selfVal instanceof ObjectValue)) {
-    		
-    		ctx.exit(this, result);
-    		return result;
-    	}
-    	
-    	MObject self = ((ObjectValue)selfVal).value();
-    	
-    	if ((isPre() && (self.state(ctx.preState()) == null)) ||
-    			(!isPre() && (self.state(ctx.postState()) == null))) {
-    		
-    		ctx.exit(this, result);
-    		return result;
-    	}
-    	
-    	MOperation operation = self.cls().operation(fOp.name(), true);
-    	
-    	if (!operation.isCallableFromOCL()) {
-    		throw new RuntimeException("Cannot call operation " + operation);
-    	}
-    	 	
-    	List<String> parameterNames = operation.paramNames();
-    	Value[] arguments = new Value[parameterNames.size()];
-    	for (int i = 1; i < fArgs.length; ++i) {
-    		arguments[i-1]= fArgs[i].eval(ctx);
-    	}
-    	
-    	// this must be done _after_ all parameters have been evaluated, 
-    	// since the parameter names could shadow values which are 
-    	// needed for a later parameter (see test\t005.*) 
-    	ctx.pushVarBinding("self", selfVal);
-    	for(int i = 0; i < parameterNames.size(); ++i) {
-    		ctx.pushVarBinding(
-    				parameterNames.get(i), 
-    				arguments[i]);
-    	}
-    	
-    	MOperationCall operationCall = 
-    		new MOperationCall(this, self, operation, arguments);
-    	
-    	operationCall.setPreferredPPCHandler(ExpressionPPCHandler.getDefaultOutputHandler());
-    	
-    	MSystem system = ctx.postState().system();
-    	
-    	try {
-    		system.enterQueryOperation(ctx, operationCall, false);
-    		operationCall.setExecutionFailed(true);
-    	
-			if (operation.hasExpression()) {
-				result = operation.expression().eval(ctx);
+		if (isPre()) {
+			ctx = new EvalContext(ctx.preState(), ctx.preState(), ctx.varBindings(), ctx);
+		}
+
+		ctx.enter(this);
+
+		Value result = UndefinedValue.instance;
+		Value selfVal = fArgs[0].eval(ctx);
+		if (selfVal.isUndefined() || !(selfVal instanceof InstanceValue)) {
+			ctx.exit(this, result);
+			return result;
+		}
+
+		MInstance self = ((InstanceValue) selfVal).value();
+		if (selfVal instanceof ObjectValue &&
+				((isPre() && (self.state(ctx.preState()) == null))
+						|| (!isPre() && (self.state(ctx.postState()) == null)))) {
+			ctx.exit(this, result);
+			return result;
+		}
+
+		if (!fOp.isCallableFromOCL()) {
```

Comment:
The search for the operation at runtime is required, because of inheritance! If `fOp` is used, overwritten operations would not be called!

Please revert to
```java
MOperation operation = self.cls().operation(fOp.name(), true);
```